File: HackerNews/Views/StoryRow.swift
import SwiftUI

/// Destinations pushed onto a `NavigationStack`.
enum Route: Hashable {
    case story(Item)
    case user(String)
}

/// One story in a list. Shared by the feeds, search results and saved items so
/// they all read the same way.
struct StoryRow: View {
    let item: Item
    var rank: Int?
    var isRead = false
    var showsSnippet = false

    @EnvironmentObject private var settings: AppSettings
    @EnvironmentObject private var library: LibraryStore

    var body: some View {
        HStack(alignment: .top, spacing: 11) {
            if let rank {
                Text("\(rank)")
                    .font(.caption.monospacedDigit())
                    .foregroundColor(.secondary)
                    .frame(minWidth: 20, alignment: .trailing)
                    .padding(.top, 3)
            } else if settings.showSourceBadges {
                SourceBadge(label: item.sourceLabel, seed: item.host ?? "ycombinator")
                    .padding(.top, 1)
            }

            VStack(alignment: .leading, spacing: Metrics.rowSpacing) {
                Text(item.displayTitle)
                    .font(.callout.weight(.semibold))
                    .foregroundColor(titleColor)
                    .lineLimit(4)
                    .multilineTextAlignment(.leading)
                    .fixedSize(horizontal: false, vertical: true)

                if let host = item.host {
                    HStack(spacing: 3) {
                        Image(systemName: "link")
                            .font(.system(size: 9))
                        Text(host)
                            .font(.caption)
                            .lineLimit(1)
                    }
                    .foregroundColor(.hnOrange)
                }

                if showsSnippet, let text = item.text, !text.isEmpty {
                    Text(HNHTML.snippet(text, limit: 160))
                        .font(.caption)
                        .foregroundColor(.secondary)
                        .lineLimit(3)
                }

                metadata
            }
        }
        .padding(.vertical, 4)
        .contentShape(Rectangle())
        .accessibilityElement(children: .combine)
        .accessibilityLabel(accessibilityLabel)
    }

    private var metadata: some View {
        HStack(spacing: 10) {
            if let score = item.score {
                MetaLabel(systemImage: "arrowtriangle.up.fill", text: score.abbreviated)
            }
            if item.kind != .job {
                MetaLabel(systemImage: "bubble.left", text: item.commentCount.abbreviated)
            }
            if let by = item.by {
                MetaLabel(systemImage: "person", text: by)
                    .lineLimit(1)
            }
            if let time = item.time {
                MetaLabel(systemImage: "clock", text: time.shortAge)
            }
            if library.isSaved(item.id) {
                Image(systemName: "bookmark.fill")
                    .font(.caption2)
                    .foregroundColor(.hnOrange)
            }
        }
        .lineLimit(1)
    }

    private var titleColor: Color {
        (isRead && settings.dimReadStories) ? .secondary : .primary
    }

    private var accessibilityLabel: String {
        var parts = [item.displayTitle]
        if let host = item.host { parts.append("from \(host)") }
        if let score = item.score { parts.append("\(score) points") }
        parts.append(item.commentCount.pluralized("comment"))
        if let by = item.by { parts.append("by \(by)") }
        if let time = item.time { parts.append(time.longAge) }
        return parts.joined(separator: ", ")
    }
}

/// Swipe actions and the long-press menu, factored out so every list that
/// shows a `StoryRow` behaves identically.
struct StoryRowActions: ViewModifier {
    let item: Item

    @EnvironmentObject private var library: LibraryStore
    @EnvironmentObject private var opener: LinkOpener

    func body(content: Content) -> some View {
        content
            .swipeActions(edge: .leading, allowsFullSwipe: true) {
                Button {
                    if library.toggleSave(item) { Haptics.success() } else { Haptics.tap() }
                } label: {
                    Label(
                        library.isSaved(item.id) ? "Unsave" : "Save",
                        systemImage: library.isSaved(item.id) ? "bookmark.slash" : "bookmark"
                    )
                }
                .tint(.hnOrange)
            }
            .swipeActions(edge: .trailing, allowsFullSwipe: false) {
                ShareLink(item: item.primaryURL) {
                    Label("Share", systemImage: "square.and.arrow.up")
                }
                .tint(.blue)

                if item.link != nil {
                    Button {
                        opener.open(item.primaryURL)
                    } label: {
                        Label("Open", systemImage: "safari")
                    }
                    .tint(.gray)
                }
            }
            .contextMenu {
                if let link = item.link {
                    Button {
                        opener.open(link)
                    } label: {
                        Label("Open Article", systemImage: "safari")
                    }
                    Button {
                        opener.openInApp(link, readerMode: true)
                    } label: {
                        Label("Open in Reader", systemImage: "doc.plaintext")
                    }
                }

                Button {
                    opener.open(item.hnURL)
                } label: {
                    Label("Open on Hacker News", systemImage: "safari")
                }

                Button {
                    if library.toggleSave(item) { Haptics.success() } else { Haptics.tap() }
                } label: {
                    Label(
                        library.isSaved(item.id) ? "Remove from Saved" : "Save Story",
                        systemImage: library.isSaved(item.id) ? "bookmark.slash" : "bookmark"
                    )
                }

                ShareLink(item: item.primaryURL) {
                    Label("Share Link", systemImage: "square.and.arrow.up")
                }
                ShareLink(item: item.hnURL) {
                    Label("Share Discussion", systemImage: "square.and.arrow.up.on.square")
                }

                Divider()

                Button {
                    UIPasteboard.general.string = item.primaryURL.absoluteString
                    Haptics.tap()
                } label: {
                    Label("Copy Link", systemImage: "doc.on.doc")
                }
            }
    }
}

extension View {
    func storyRowActions(for item: Item) -> some View {
        modifier(StoryRowActions(item: item))
    }
}

File: HackerNewsAndroid/app/src/main/java/com/nurtrino/hackernews/ui/StoryRow.kt
package com.nurtrino.hackernews.ui

import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material.icons.filled.DateRange
import androidx.compose.material.icons.filled.List
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.nurtrino.hackernews.data.Library
import com.nurtrino.hackernews.data.Settings
import com.nurtrino.hackernews.model.Item
import com.nurtrino.hackernews.model.ItemKind
import com.nurtrino.hackernews.text.HnHtml

/**
 * One story in a list. Shared by the feeds, search results and saved items so
 * they all read the same way.
 */
@Composable
fun StoryRow(
    item: Item,
    settings: Settings,
    library: Library,
    rank: Int? = null,
    showSnippet: Boolean = false,
    onClick: () -> Unit,
) {
    val context = LocalContext.current
    var menuOpen by remember { mutableStateOf(false) }
    val isRead = library.isRead(item.id)
    val saved = library.isSaved(item.id)

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .combinedClickable(onClick = onClick, onLongClick = { menuOpen = true })
            .padding(horizontal = 16.dp, vertical = 10.dp),
        verticalAlignment = Alignment.Top,
    ) {
        if (rank != null) {
            Text(
                rank.toString(),
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.widthIn(min = 22.dp).padding(top = 2.dp, end = 10.dp),
            )
        } else if (settings.showSourceBadges) {
            Row(modifier = Modifier.padding(end = 11.dp)) {
                SourceBadge(item.sourceLabel, item.host ?: "ycombinator")
            }
        }

        Column(modifier = Modifier.fillMaxWidth()) {
            Text(
                item.displayTitle,
                style = MaterialTheme.typography.bodyLarge,
                fontWeight = FontWeight.SemiBold,
                color = if (isRead && settings.dimReadStories) {
                    MaterialTheme.colorScheme.onSurfaceVariant
                } else {
                    MaterialTheme.colorScheme.onSurface
                },
                maxLines = 4,
            )

            item.host?.let { host ->
                Text(
                    host,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.primary,
                    maxLines = 1,
                    modifier = Modifier.padding(top = 4.dp),
                )
            }

            if (showSnippet && !item.text.isNullOrEmpty()) {
                Text(
                    HnHtml.snippet(item.text, 160),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    maxLines = 3,
                    modifier = Modifier.padding(top = 4.dp),
                )
            }

            Row(
                modifier = Modifier.padding(top = 6.dp),
                horizontalArrangement = Arrangement.spacedBy(10.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                item.score?.let { MetaLabel(Icons.Filled.KeyboardArrowUp, abbreviated(it)) }
                if (item.kind != ItemKind.JOB) {
                    MetaLabel(Icons.Filled.List, abbreviated(item.commentCount))
                }
                item.by?.let { MetaLabel(Icons.Filled.Person, it) }
                MetaLabel(Icons.Filled.DateRange, shortAge(item.time))
                if (saved) {
                    Icon(
                        Icons.Filled.Star,
                        contentDescription = "Saved",
                        modifier = Modifier.size(14.dp),
                        tint = MaterialTheme.colorScheme.primary,
                    )
                }
            }
        }
    }

    DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
        item.link?.let { link ->
            DropdownMenuItem(
                text = { Text("Open Article") },
                onClick = {
                    menuOpen = false
                    openLink(context, link, settings.linkTarget)
                },
            )
        }
        DropdownMenuItem(
            text = { Text("Open on Hacker News") },
            onClick = {
                menuOpen = false
                openLink(context, item.hnUrl, settings.linkTarget)
            },
        )
        DropdownMenuItem(
            text = { Text(if (saved) "Remove from Saved" else "Save Story") },
            onClick = {
                menuOpen = false
                library.toggleSave(item)
            },
        )
        HorizontalDivider()
        DropdownMenuItem(
            text = { Text("Share Link") },
            onClick = {
                menuOpen = false
                shareLink(context, item.primaryUrl, item.displayTitle)
            },
        )
        DropdownMenuItem(
            text = { Text("Share Discussion") },
            onClick = {
                menuOpen = false
                shareLink(context, item.hnUrl, item.displayTitle)
            },
        )
        DropdownMenuItem(
            text = { Text("Copy Link") },
            onClick = {
                menuOpen = false
                copyToClipboard(context, item.primaryUrl)
            },
        )
    }
}
